File: example_sleep_analysis.R
# Example: Age Stratification in Sleep Research
# This script demonstrates how to apply age stratification to sleep and circadian data

# Load the age stratification functions
source("age_stratification.R")

# Load required libraries (install if needed)
# install.packages(c("dplyr", "ggplot2"))
library(dplyr)
library(ggplot2)

# ============================================================================
# STEP 1: Create example sleep and circadian data
# ============================================================================

set.seed(42)
n <- 100

# Create synthetic dataset representing sleep and circadian measures
sleep_data <- data.frame(
  participant_id = 1:n,
  age = round(runif(n, 40, 75)),
  
  # Sleep measures (with age-related trends)
  sleep_efficiency = pmax(50, pmin(100, 88 - 0.15 * (runif(n, 40, 75) - 40) + rnorm(n, 0, 8))),
  total_sleep_time = pmax(300, 420 - 0.8 * (runif(n, 40, 75) - 40) + rnorm(n, 0, 45)),
  waso = pmax(0, 25 + 0.5 * (runif(n, 40, 75) - 40) + abs(rnorm(n, 0, 15))),
  
  # Circadian measures
  circadian_phase = 2.5 + 0.02 * (runif(n, 40, 75) - 40) + rnorm(n, 0, 1.2),
  amplitude = pmax(0.1, 0.8 - 0.006 * (runif(n, 40, 75) - 40) + abs(rnorm(n, 0, 0.15))),
  
  # Biological measures
  fasting_glucose = pmax(70, 92 + 0.3 * (runif(n, 40, 75) - 40) + rnorm(n, 0, 10)),
  cortisol_awakening = pmax(5, 18 - 0.08 * (runif(n, 40, 75) - 40) + rnorm(n, 0, 4))
)

# ============================================================================
# STEP 2: Apply different age stratification methods
# ============================================================================

cat("Age Stratification Analysis for Sleep Research\n")
cat("===============================================\n\n")

# Method 1: Quartiles (recommended for balanced analysis)
sleep_data$age_quartile <- stratify_age_quartiles(sleep_data$age)

# Method 2: Decades (recommended for communication)
sleep_data$age_decade <- stratify_age_decades(sleep_data$age)

# Method 3: Custom groups (example for metabolic research)
sleep_data$age_custom <- stratify_age_custom(sleep_data$age, 
                                              breaks = c(40, 50, 60, 70, 75))

# ============================================================================
# STEP 3: Analyze sleep efficiency by age group (using quartiles)
# ============================================================================

cat("Sleep Efficiency by Age Quartile\n")
cat("---------------------------------\n")

sleep_summary <- sleep_data %>%
  group_by(age_quartile) %>%
  summarise(
    N = n(),
    Mean_Age = round(mean(age), 1),
    SD_Age = round(sd(age), 1),
    Mean_Sleep_Eff = round(mean(sleep_efficiency), 1),
    SD_Sleep_Eff = round(sd(sleep_efficiency), 1),
    Mean_TST = round(mean(total_sleep_time), 0),
    Mean_WASO = round(mean(waso), 1)
  )

print(sleep_summary)
cat("\n")

# Statistical test
kruskal_result <- kruskal.test(sleep_efficiency ~ age_quartile, data = sleep_data)
cat(sprintf("Kruskal-Wallis test: H = %.2f, p = %.4f\n\n", 
            kruskal_result$statistic, kruskal_result$p.value))

# ============================================================================
# STEP 4: Analyze circadian phase by age decade
# ============================================================================

cat("Circadian Phase by Age Decade\n")
cat("------------------------------\n")

circadian_summary <- sleep_data %>%
  group_by(age_decade) %>%
  summarise(
    N = n(),
    Mean_Age = round(mean(age), 1),
    Mean_Phase = round(mean(circadian_phase), 2),
    SD_Phase = round(sd(circadian_phase), 2),
    Mean_Amplitude = round(mean(amplitude), 2)
  )

print(circadian_summary)
cat("\n")

# ============================================================================
# STEP 5: Analyze glucose by custom age groups
# ============================================================================

cat("Fasting Glucose by Custom Age Groups\n")
cat("-------------------------------------\n")

glucose_summary <- sleep_data %>%
  group_by(age_custom) %>%
  summarise(
    N = n(),
    Mean_Age = round(mean(age), 1),
    Mean_Glucose = round(mean(fasting_glucose), 1),
    SD_Glucose = round(sd(fasting_glucose), 1),
    Above_100 = sum(fasting_glucose >= 100),
    Pct_Above_100 = round(100 * sum(fasting_glucose >= 100) / n(), 1)
  )

print(glucose_summary)
cat("\n")

# ============================================================================
# STEP 6: Create visualizations
# ============================================================================

cat("Creating visualizations...\n")

# Plot 1: Sleep efficiency by age quartile
p1 <- ggplot(sleep_data, aes(x = age_quartile, y = sleep_efficiency)) +
  geom_boxplot(fill = "lightblue", alpha = 0.7) +
  geom_jitter(width = 0.2, alpha = 0.3) +
  labs(title = "Sleep Efficiency by Age Quartile",
       x = "Age Quartile",
       y = "Sleep Efficiency (%)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Plot 2: Circadian phase by age (with decade coloring)
p2 <- ggplot(sleep_data, aes(x = age, y = circadian_phase, color = age_decade)) +
  geom_point(alpha = 0.6, size = 2) +
  geom_smooth(method = "lm", se = TRUE, color = "black", linetype = "dashed") +
  labs(title = "Circadian Phase vs Age",
       x = "Age (years)",
       y = "Circadian Phase (hours)",
       color = "Age Decade") +
  theme_minimal()

# Plot 3: Multiple measures by age quartile
sleep_long <- sleep_data %>%
  select(age_quartile, sleep_efficiency, total_sleep_time, waso) %>%
  tidyr::pivot_longer(cols = c(sleep_efficiency, total_sleep_time, waso),
                     names_to = "measure",
                     values_to = "value")

p3 <- ggplot(sleep_long, aes(x = age_quartile, y = value)) +
  geom_boxplot(fill = "lightgreen", alpha = 0.7) +
  facet_wrap(~ measure, scales = "free_y") +
  labs(title = "Sleep Measures by Age Quartile",
       x = "Age Quartile",
       y = "Value") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Save plots
ggsave("sleep_efficiency_by_age.png", p1, width = 8, height = 6)
ggsave("circadian_phase_vs_age.png", p2, width = 8, height = 6)
ggsave("multiple_sleep_measures.png", p3, width = 10, height = 6)

cat("Plots saved successfully!\n\n")

# ============================================================================
# STEP 7: Export stratified data
# ============================================================================

cat("Exporting stratified data...\n")

# Save data with age stratifications
write.csv(sleep_data, "sleep_data_with_age_groups.csv", row.names = FALSE)

cat("Data exported to: sleep_data_with_age_groups.csv\n\n")

# ============================================================================
# STEP 8: Summary report
# ============================================================================

cat("Summary Report\n")
cat("==============\n")
cat(sprintf("Total participants: %d\n", nrow(sleep_data)))
cat(sprintf("Age range: %d - %d years\n", min(sleep_data$age), max(sleep_data$age)))
cat(sprintf("Mean age: %.1f (SD = %.1f)\n\n", mean(sleep_data$age), sd(sleep_data$age)))

cat("Age stratification methods applied:\n")
cat("1. Quartiles - for balanced statistical analysis\n")
cat("2. Decades - for easy interpretation and reporting\n")
cat("3. Custom groups (40-49, 50-59, 60-69, 70-74) - for targeted analysis\n\n")

cat("Key findings:\n")
cat(sprintf("- Sleep efficiency decreases with age (Q1: %.1f%%, Q4: %.1f%%)\n",
            mean(sleep_data$sleep_efficiency[sleep_data$age_quartile == "Q1 (Youngest 25%)"]),
            mean(sleep_data$sleep_efficiency[sleep_data$age_quartile == "Q4 (Oldest 25%)"])))

cat(sprintf("- Wake after sleep onset increases with age (Q1: %.1f min, Q4: %.1f min)\n",
            mean(sleep_data$waso[sleep_data$age_quartile == "Q1 (Youngest 25%)"]),
            mean(sleep_data$waso[sleep_data$age_quartile == "Q4 (Oldest 25%)"])))

cat(sprintf("- Fasting glucose increases with age (40s: %.1f mg/dL, 70s: %.1f mg/dL)\n",
            mean(sleep_data$fasting_glucose[sleep_data$age_decade == "40s (40-49)"]),
            mean(sleep_data$fasting_glucose[sleep_data$age_decade == "70s (70-79)"])))

cat("\nAnalysis complete!\n")

File: example_sleep_analysis.py
"""
Example: Age Stratification in Sleep Research
This script demonstrates how to apply age stratification to sleep and circadian data
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from age_stratification import (
    stratify_age_quartiles,
    stratify_age_decades,
    stratify_age_custom
)

# Set style for plots
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (10, 6)

# ============================================================================
# STEP 1: Create example sleep and circadian data
# ============================================================================

np.random.seed(42)
n = 100

# Create synthetic dataset representing sleep and circadian measures
ages = np.random.uniform(40, 75, n)

sleep_data = pd.DataFrame({
    'participant_id': range(1, n+1),
    'age': np.round(ages),
    
    # Sleep measures (with age-related trends)
    'sleep_efficiency': np.clip(88 - 0.15 * (ages - 40) + np.random.normal(0, 8, n), 50, 100),
    'total_sleep_time': np.clip(420 - 0.8 * (ages - 40) + np.random.normal(0, 45, n), 300, 600),
    'waso': np.clip(25 + 0.5 * (ages - 40) + np.abs(np.random.normal(0, 15, n)), 0, 120),
    
    # Circadian measures
    'circadian_phase': 2.5 + 0.02 * (ages - 40) + np.random.normal(0, 1.2, n),
    'amplitude': np.clip(0.8 - 0.006 * (ages - 40) + np.abs(np.random.normal(0, 0.15, n)), 0.1, 2),
    
    # Biological measures
    'fasting_glucose': np.clip(92 + 0.3 * (ages - 40) + np.random.normal(0, 10, n), 70, 150),
    'cortisol_awakening': np.clip(18 - 0.08 * (ages - 40) + np.random.normal(0, 4, n), 5, 35)
})

# ============================================================================
# STEP 2: Apply different age stratification methods
# ============================================================================

print("Age Stratification Analysis for Sleep Research")
print("=" * 60)
print()

# Method 1: Quartiles (recommended for balanced analysis)
sleep_data['age_quartile'] = stratify_age_quartiles(sleep_data['age'])

# Method 2: Decades (recommended for communication)
sleep_data['age_decade'] = stratify_age_decades(sleep_data['age'])

# Method 3: Custom groups (example for metabolic research)
sleep_data['age_custom'] = stratify_age_custom(sleep_data['age'], 
                                                breaks=[40, 50, 60, 70, 75])

# ============================================================================
# STEP 3: Analyze sleep efficiency by age group (using quartiles)
# ============================================================================

print("Sleep Efficiency by Age Quartile")
print("-" * 60)

sleep_summary = sleep_data.groupby('age_quartile').agg({
    'age': ['count', 'mean', 'std'],
    'sleep_efficiency': ['mean', 'std'],
    'total_sleep_time': 'mean',
    'waso': 'mean'
}).round(1)

print(sleep_summary)
print()

# Statistical test
groups = [group['sleep_efficiency'].values 
          for name, group in sleep_data.groupby('age_quartile')]
h_stat, p_value = stats.kruskal(*groups)
print(f"Kruskal-Wallis test: H = {h_stat:.2f}, p = {p_value:.4f}")
print()

# ============================================================================
# STEP 4: Analyze circadian phase by age decade
# ============================================================================

print("Circadian Phase by Age Decade")
print("-" * 60)

circadian_summary = sleep_data.groupby('age_decade').agg({
    'age': ['count', 'mean'],
    'circadian_phase': ['mean', 'std'],
    'amplitude': 'mean'
}).round(2)

print(circadian_summary)
print()

# ============================================================================
# STEP 5: Analyze glucose by custom age groups
# ============================================================================

print("Fasting Glucose by Custom Age Groups")
print("-" * 60)

glucose_summary = sleep_data.groupby('age_custom').apply(lambda x: pd.Series({
    'N': len(x),
    'Mean_Age': x['age'].mean(),
    'Mean_Glucose': x['fasting_glucose'].mean(),
    'SD_Glucose': x['fasting_glucose'].std(),
    'Above_100': (x['fasting_glucose'] >= 100).sum(),
    'Pct_Above_100': 100 * (x['fasting_glucose'] >= 100).sum() / len(x)
})).round(1)

print(glucose_summary)
print()

# ============================================================================
# STEP 6: Create visualizations
# ============================================================================

print("Creating visualizations...")

# Plot 1: Sleep efficiency by age quartile
fig, ax = plt.subplots(figsize=(10, 6))
sleep_data.boxplot(column='sleep_efficiency', by='age_quartile', ax=ax)
ax.set_xlabel('Age Quartile')
ax.set_ylabel('Sleep Efficiency (%)')
ax.set_title('Sleep Efficiency by Age Quartile')
plt.suptitle('')  # Remove the default title
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.savefig('sleep_efficiency_by_age_python.png', dpi=300, bbox_inches='tight')
plt.close()

# Plot 2: Circadian phase vs age (with decade coloring)
fig, ax = plt.subplots(figsize=(10, 6))
for decade in sleep_data['age_decade'].unique():
    mask = sleep_data['age_decade'] == decade
    ax.scatter(sleep_data[mask]['age'], 
              sleep_data[mask]['circadian_phase'],
              label=decade, alpha=0.6, s=50)

# Add trend line
z = np.polyfit(sleep_data['age'], sleep_data['circadian_phase'], 1)
p = np.poly1d(z)
ax.plot(sleep_data['age'].sort_values(), 
        p(sleep_data['age'].sort_values()), 
        "k--", alpha=0.5, label='Trend')

ax.set_xlabel('Age (years)')
ax.set_ylabel('Circadian Phase (hours)')
ax.set_title('Circadian Phase vs Age')
ax.legend()
plt.tight_layout()
plt.savefig('circadian_phase_vs_age_python.png', dpi=300, bbox_inches='tight')
plt.close()

# Plot 3: Multiple measures by age quartile
fig, axes = plt.subplots(1, 3, figsize=(15, 5))

measures = ['sleep_efficiency', 'total_sleep_time', 'waso']
titles = ['Sleep Efficiency', 'Total Sleep Time', 'WASO']

for ax, measure, title in zip(axes, measures, titles):
    sleep_data.boxplot(column=measure, by='age_quartile', ax=ax)
    ax.set_xlabel('Age Quartile')
    ax.set_ylabel(title)
    ax.set_title(title)
    plt.sca(ax)
    plt.xticks(rotation=45, ha='right')

plt.suptitle('Sleep Measures by Age Quartile')
plt.tight_layout()
plt.savefig('multiple_sleep_measures_python.png', dpi=300, bbox_inches='tight')
plt.close()

# Plot 4: Heatmap of correlations
fig, ax = plt.subplots(figsize=(10, 8))
corr_data = sleep_data[['age', 'sleep_efficiency', 'total_sleep_time', 
                        'waso', 'circadian_phase', 'amplitude', 
                        'fasting_glucose', 'cortisol_awakening']]
correlation = corr_data.corr()
sns.heatmap(correlation, annot=True, cmap='coolwarm', center=0, ax=ax)
ax.set_title('Correlation Matrix of Age and Sleep/Circadian Measures')
plt.tight_layout()
plt.savefig('correlation_heatmap_python.png', dpi=300, bbox_inches='tight')
plt.close()

print("Plots saved successfully!")
print()

# ============================================================================
# STEP 7: Export stratified data
# ============================================================================

print("Exporting stratified data...")

# Save data with age stratifications
sleep_data.to_csv('sleep_data_with_age_groups_python.csv', index=False)

print("Data exported to: sleep_data_with_age_groups_python.csv")
print()

# ============================================================================
# STEP 8: Summary report
# ============================================================================

print("Summary Report")
print("=" * 60)
print(f"Total participants: {len(sleep_data)}")
print(f"Age range: {int(sleep_data['age'].min())} - {int(sleep_data['age'].max())} years")
print(f"Mean age: {sleep_data['age'].mean():.1f} (SD = {sleep_data['age'].std():.1f})")
print()

print("Age stratification methods applied:")
print("1. Quartiles - for balanced statistical analysis")
print("2. Decades - for easy interpretation and reporting")
print("3. Custom groups (40-49, 50-59, 60-69, 70-74) - for targeted analysis")
print()

# Calculate key findings
q1_data = sleep_data[sleep_data['age_quartile'] == 'Q1 (Youngest 25%)']
q4_data = sleep_data[sleep_data['age_quartile'] == 'Q4 (Oldest 25%)']
decade_40s = sleep_data[sleep_data['age_decade'] == '40s (40-49)']
decade_70s = sleep_data[sleep_data['age_decade'] == '70s (70-79)']

print("Key findings:")
print(f"- Sleep efficiency decreases with age (Q1: {q1_data['sleep_efficiency'].mean():.1f}%, "
      f"Q4: {q4_data['sleep_efficiency'].mean():.1f}%)")
print(f"- Wake after sleep onset increases with age (Q1: {q1_data['waso'].mean():.1f} min, "
      f"Q4: {q4_data['waso'].mean():.1f} min)")

if len(decade_70s) > 0:
    print(f"- Fasting glucose increases with age (40s: {decade_40s['fasting_glucose'].mean():.1f} mg/dL, "
          f"70s: {decade_70s['fasting_glucose'].mean():.1f} mg/dL)")

print()
print("Analysis complete!")
